File: pyg4ometry/pygeometry/cpgeant4/solid/Trd.cpp
#include "Trd.h"

CSG* CSGMesh::ConstructTrd(double pX1,double pX2,double pY1,double pY2,double pZ){
  std::vector<Polygon*> polygons(6);

  std::vector<Vertex*> vtx1(4);
  vtx1[0] = new Vertex(new Vector(-pX1,-pY1,-pZ), NULL);
  vtx1[1] = new Vertex(new Vector(-pX1, pY1,-pZ), NULL);
  vtx1[2] = new Vertex(new Vector( pX1, pY1,-pZ), NULL);
  vtx1[3] = new Vertex(new Vector( pX1,-pY1,-pZ), NULL);
  polygons[0] = new Polygon(vtx1,NULL);

  std::vector<Vertex*> vtx2(4);
  vtx2[0] = new  Vertex(new Vector(-pX2,-pY2, pZ), NULL);
  vtx2[1] = new  Vertex(new Vector( pX2,-pY2, pZ), NULL);
  vtx2[2] = new  Vertex(new Vector( pX2, pY2, pZ), NULL);
  vtx2[3] = new  Vertex(new Vector(-pX2, pY2, pZ), NULL);
  polygons[1] = new Polygon(vtx2,NULL);

  std::vector<Vertex*> vtx3(4);
  vtx3[0] = new  Vertex(new Vector(-pX1,-pY1,-pZ), NULL);
  vtx3[1] = new  Vertex(new Vector( pX1,-pY1,-pZ), NULL);
  vtx3[2] = new  Vertex(new Vector( pX2,-pY2, pZ), NULL);
  vtx3[3] = new  Vertex(new Vector(-pX2,-pY2, pZ), NULL);
  polygons[2] = new Polygon(vtx3,NULL);

  std::vector<Vertex*> vtx4(4);
  vtx4[0] = new  Vertex(new Vector(-pX1, pY1,-pZ), NULL);
  vtx4[1] = new  Vertex(new Vector(-pX2, pY2, pZ), NULL);
  vtx4[2] = new  Vertex(new Vector( pX2, pY2, pZ), NULL);
  vtx4[3] = new  Vertex(new Vector( pX1, pY1,-pZ), NULL);
  polygons[3] = new Polygon(vtx4,NULL);

  std::vector<Vertex*> vtx5(4);
  vtx5[0] = new  Vertex(new Vector(-pX1,-pY1,-pZ), NULL);
  vtx5[1] = new  Vertex(new Vector(-pX2,-pY2, pZ), NULL);
  vtx5[2] = new  Vertex(new Vector(-pX2, pY2, pZ), NULL);
  vtx5[3] = new  Vertex(new Vector(-pX1, pY1,-pZ), NULL);
  polygons[4] = new Polygon(vtx5,NULL);

  std::vector<Vertex*> vtx6(4);
  vtx6[0] = new  Vertex(new Vector( pX1,-pY1,-pZ), NULL);
  vtx6[1] = new  Vertex(new Vector( pX1, pY1,-pZ), NULL);
  vtx6[2] = new  Vertex(new Vector( pX2, pY2, pZ), NULL);
  vtx6[3] = new  Vertex(new Vector( pX2,-pY2, pZ), NULL);
  polygons[5] = new Polygon(vtx6,NULL);

  return CSG::fromPolygons(polygons);
}

File: pyg4ometry/geant4/solid/Trd.py
from ... import config as _config

from .SolidBase import SolidBase as _SolidBase

if _config.meshing == _config.meshingType.pycsg :
    from pyg4ometry.pycsg.core import CSG as _CSG
    from pyg4ometry.pycsg.geom import Vector as _Vector
    from pyg4ometry.pycsg.geom import Vertex as _Vertex
    from pyg4ometry.pycsg.geom import Polygon as _Polygon
elif _config.meshing == _config.meshingType.cgal_sm :
    from pyg4ometry.pycgal.core import CSG as _CSG
    from pyg4ometry.pycgal.geom import Vector as _Vector
    from pyg4ometry.pycgal.geom import Vertex as _Vertex
    from pyg4ometry.pycgal.geom import Polygon as _Polygon

import logging as _log

class Trd(_SolidBase):
    """
    Constructs a trapezoid.

    :param name:  of the solid
    :type name:   str
    :param pDx1:  length along x at the surface positioned at -dz/2
    :type pDx1:   float, Constant, Quantity, Variable
    :param pDx2:  length along x at the surface postitioned at +dz/2
    :type pDx2:   float, Constant, Quantity, Variable
    :param pDy1:  length along y at the surface positioned at -dz/2
    :type pDy1:   float, Constant, Quantity, Variable
    :param pDy2:  length along y at the surface positioned at +dz/2
    :type pDy2:   float, Constant, Quantity, Variable
    :param dz:    length along the z axis
    :type dz:     float, Constant, Quantity, Variable
    :param registry: for storing solid
    :type registry: Registry
    :param lunit: length unit (nm,um,mm,m,km) for solid
    :type lunit: str
    """
    def __init__(self, name, pDx1, pDx2, pDy1, pDy2, pDz, registry, lunit="mm", addRegistry=True):
        super(Trd, self).__init__(name, 'Trd', registry)

        self.pX1    = pDx1
        self.pX2    = pDx2
        self.pY1    = pDy1
        self.pY2    = pDy2
        self.pZ     = pDz
        self.lunit  = lunit

        self.dependents = []

        self.varNames = ["pX1", "pX2", "pY1", "pY2", "pZ"]
        self.varUnits = ["lunit", "lunit", "lunit", "lunit", "lunit"]

        if addRegistry:
            registry.addSolid(self)

    def mesh(self):
        _log.info('trd.pycsgmesh> antlr')
        import pyg4ometry.gdml.Units as _Units #TODO move circular import 
        luval = _Units.unit(self.lunit)

        pX1 = self.evaluateParameter(self.pX1)*luval/2.
        pX2 = self.evaluateParameter(self.pX2)*luval/2.
        pY1 = self.evaluateParameter(self.pY1)*luval/2.
        pY2 = self.evaluateParameter(self.pY2)*luval/2.
        pZ  = self.evaluateParameter(self.pZ)*luval/2.

        _log.info('trd.pycsgmesh> mesh')
        mesh  = _CSG.fromPolygons([_Polygon([_Vertex(_Vector(-pX2,  pY2, pZ)),
                                             _Vertex(_Vector(-pX2, -pY2, pZ)),
                                             _Vertex(_Vector(pX2,  -pY2, pZ)),
                                             _Vertex(_Vector(pX2,  pY2,  pZ))]),
                                   
                                   _Polygon([ _Vertex(_Vector(-pX1, -pY1, -pZ)),
                                              _Vertex(_Vector(-pX1,  pY1, -pZ)),
                                              _Vertex(_Vector(pX1,   pY1, -pZ)),
                                              _Vertex(_Vector(pX1,  -pY1, -pZ))]),
                                   
                                   _Polygon([_Vertex(_Vector(pX2,  -pY2,  pZ)),
                                             _Vertex(_Vector(-pX2, -pY2,  pZ)),
                                             _Vertex(_Vector(-pX1, -pY1, -pZ)),
                                             _Vertex(_Vector(pX1,  -pY1, -pZ))]),
                                   
                                   _Polygon([_Vertex(_Vector(pX2,  pY2,  pZ)),
                                             _Vertex(_Vector(pX1,  pY1, -pZ)),
                                             _Vertex(_Vector(-pX1, pY1, -pZ)),
                                             _Vertex(_Vector(-pX2, pY2,  pZ))]),
                                   
                                   _Polygon([_Vertex(_Vector(-pX2,  pY2,  pZ)),
                                             _Vertex(_Vector(-pX1,  pY1, -pZ)),
                                             _Vertex(_Vector(-pX1, -pY1, -pZ)),
                                             _Vertex(_Vector(-pX2, -pY2,  pZ))]),
                                   
                                   _Polygon([_Vertex(_Vector(pX2, -pY2,  pZ)),
                                             _Vertex(_Vector(pX1, -pY1, -pZ)),
                                             _Vertex(_Vector(pX1,  pY1, -pZ)),
                                             _Vertex(_Vector(pX2,  pY2,  pZ))])])
        
        return mesh

